clang/lib/Sema/SemaBounds.cpp

```diff
@@ -605,14 +605,21 @@ namespace {
 
       return true;
     }
+
+    // Ensure that any variables within bounds value expressions are detected
+    // by explicitly traversing the child of a bounds value expression.
+    bool VisitBoundsValueExpr(BoundsValueExpr *E) {
+      TraverseStmt(E->getTemporaryBinding());
```

Comment:
Traversing the temporary binding of a `BoundsValueExpr` by default in `RecursiveASTVisitor` breaks non-modifying expression detection in inferred bounds. Example from checkedc/tests/typechecking/pointer_types.c:

```
ptr<int> t2 = h1();
```

error: inferred bounds 'bounds(value of h1(), value of h1() + 1)' contain a modifying expression; use a temporary instead